chore: Make all applicable spec links use the latest spec instead of a specific version

zecakeh opened the pull request:
Will avoid having to update the Matrix version in the spec links after each release.













<!-- Replace -->
----
Preview: https://pr-1473--ruma-docs.surge.sh
<!-- Replace -->

zecakeh committed:
xtask: Allow spec links to use the `latest` spec instead of a specific version

zecakeh committed:
chore: Make all applicable spec links use the latest spec instead of a specific version

zecakeh committed:
xtask: Disallow linking to the unstable spec

This can be re-enabled if a legitimate use case is found.

ShadowJonathan commented:
The reason we haven't done this before is because of https://github.com/ruma/ruma/issues/871; i'd like first for `xtask ci` to check wether or not the spec link fragments are dangling. Putting it on `/latest/` may break it in the future, and i'd like us to be proactive with this instead of reactive when some small documentation bugs come in, it'd also damage the reputation of ruma somewhat if they experience those small link breakages, i'd rather us check it automatically than not, it's a small effort.

zecakeh commented:
I could look into it if you want. Since the issue is assigned to you I didn't want to overstep.

ShadowJonathan commented:
Ah, that may just be because of a todo item thing i've had left.

Considering the workload uni is throwing at me now, and how long ive had myself assigned to it, i'll just unassign myself.

zecakeh commented:
Done in #1474. It should probably be merged before this so we can check nothing is broken here.

zecakeh commented:
Rebased on `main` with the new spec links check.

jplatte commented on crates/ruma-common/src/push/condition.rs:
```diff
@@ -590,7 +590,7 @@ mod tests {
         assert!("[[:alpha:]]!".matches_word("[[:alpha:]]?"));
 
-        // From the spec: <https://spec.matrix.org/v1.4/client-server-api/#conditions-1>
+        // From the spec: <https://spec.matrix.org/latest/client-server-api/#conditions-1>
```
Maybe the changes in this file should be reverted? There seems to be a high-ish risk that this link will refer to the wrong section in the future (though it seems correct right now).

zecakeh replied:
We could add a condition in the spec links check that fragments ending with a number shouldn't use the `latest` version?

ShadowJonathan replied:
The ones without effectively end with `-0`, so I don't think that's a good idea

jplatte replied:
We could go for "if there's any duplicates that would be assigned `-num`, don't allow `/latest/` link" I guess. But I'd also be fine with not automating these exceptions.

zecakeh replied:
> The ones without effectively end with -0, so I don't think that's a good idea

No they don't, there's never a `-0` suffix. The first one doesn't have a number at the end, the second one has `-1` and so on. (This also means that any link can point to the wrong fragment if the same ID is added before at some point.)

> But I'd also be fine with not automating these exceptions.

I'd rather automate it. Off the top of my head I remember having 5 invalid links ending with a number before I handled it, we only have 3 here :slightly_smiling_face: (I might be wrong though).

Edit: I confirm there are 5, but one of them is unlikely to change.

ShadowJonathan replied:
Tbh I'd rather petition the spec to start using header ordering numbers (1.4.6. Etc) than go through with this change tbh, there's just too much risk of silent "breakage"

jplatte replied:
> This also means that any link can point to the wrong fragment if the same ID is added before at some point.

I think that's what Jonathan meant with there being "effectively" a `-0` at the end for links that don't otherwise have a number.

> I'd rather automate it.

Do you think my suggestion is realistic there? I.e. raise an error on a link to `/latest/...#foo` if `foo-1` exists?

zecakeh replied:
> Do you think my suggestion is realistic there? I.e. raise an error on a link to /latest/...#foo if foo-1 exists?

Yes, no problem

jplatte commented on crates/ruma-identifiers-validation/src/error.rs:
```diff
@@ -77,7 +77,7 @@ pub enum MxcUriError {
     ///
     /// Valid characters are (in regex notation) `[A-Za-z0-9_-]+`.
-    /// See [here](https://spec.matrix.org/v1.4/client-server-api/#security-considerations-5) for more details.
+    /// See [here](https://spec.matrix.org/latest/client-server-api/#security-considerations-5) for more details.
```
Same here, I see a very high chance this link will refer to the wrong place in the future.

jplatte commented on crates/ruma-identifiers-validation/src/mxc_uri.rs:
```diff
@@ -17,7 +17,7 @@ pub fn validate(uri: &str) -> Result<NonZeroU8, MxcUriError> {
     let server_name = &uri[..index];
     let media_id = &uri[index + 1..];
-    // See: https://spec.matrix.org/v1.4/client-server-api/#security-considerations-5
+    // See: https://spec.matrix.org/latest/client-server-api/#security-considerations-5
```
Same link.

zecakeh committed:
xtask: Disallow linking to the latest spec when a link uses a non-unique ID

These IDs can change from one version to the other if another heading
with the same name is added before, so the link would not
point to the proper part of the spec anymore.

jplatte commented on xtask/src/ci/spec_links.rs:
```diff
@@ -72,6 +72,16 @@ impl SpecLink {
+    /// The ID has duplicates.
+    Yes,
+    /// The ID doesn't have duplicates.
+    #[default]
```
The `Default` impl doesn't seem to be used, or am I missing sth.?

zecakeh replied:
It isn't, I thought I might need it :slightly_smiling_face:. I _could_ use it instead of using `No`.
